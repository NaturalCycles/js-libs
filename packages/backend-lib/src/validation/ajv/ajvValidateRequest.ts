import type { AjvSchema, AjvValidationError } from '@naturalcycles/nodejs-lib/ajv'
import type { BackendRequest } from '../../server/server.model.js'
import { handleValidationError, type ReqValidationOptions } from '../validateRequest.util.js'

class AjvValidateRequest {
  body<T>(
    req: BackendRequest,
    schema: AjvSchema<T>,
    opt: ReqValidationOptions<AjvValidationError> = {},
  ): T {
    return this.validate(req, 'body', schema, opt)
  }

  query<T>(
    req: BackendRequest,
    schema: AjvSchema<T>,
    opt: ReqValidationOptions<AjvValidationError> = {},
  ): T {
    return this.validate(req, 'query', schema, opt)
  }

  params<T>(
    req: BackendRequest,
    schema: AjvSchema<T>,
    opt: ReqValidationOptions<AjvValidationError> = {},
  ): T {
    return this.validate(req, 'params', schema, opt)
  }

  /**
   * Validates `req.headers` against the provided schema.
   *
   * Note: as opposed to other methods, this method does not mutate `req.headers` in case of success,
   * i.e. schemas that cast values will not have any effect.
   *
   * If you wish to mutate `req.headers` with the validated value, use `keepOriginal: false` option.
   * Keep in mind that this will also remove all values that are not in the schema.
   */
  headers<T>(
    req: BackendRequest,
    schema: AjvSchema<T>,
    opt: ReqValidationOptions<AjvValidationError> = {},
  ): T {
    return this.validate(req, 'headers', schema, opt)
  }

  private validate<T>(
    req: BackendRequest,
    reqProperty: 'body' | 'params' | 'query' | 'headers',
    schema: AjvSchema<T>,
    opt: ReqValidationOptions<AjvValidationError> = {},
  ): T {
    const { mutate } = opt
    const originalProperty = req[reqProperty] || {}
    const item: T = mutate ? originalProperty : { ...originalProperty }

    // Ajv mutates the input
    const error = schema.getValidationError(item, {
      objectName: `request ${reqProperty}`,
    })

    if (error) {
      handleValidationError(error, originalProperty, opt)
    }

    return item
  }
}

export const ajvValidateRequest = new AjvValidateRequest()
