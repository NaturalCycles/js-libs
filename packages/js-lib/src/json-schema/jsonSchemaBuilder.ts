import { _uniq } from '../array/array.util.js'
import { _numberEnumValues, _stringEnumValues, isNumberEnum, isStringEnum } from '../enum.util.js'
import { _deepCopy } from '../object/object.util.js'
import { _sortObject } from '../object/sortObject.js'
import {
  type AnyObject,
  type BaseDBEntity,
  type IsoDate,
  type IsoDateTime,
  JWT_REGEX,
  type NumberEnum,
  type StringEnum,
  type UnixTimestamp,
  type UnixTimestampMillis,
} from '../types.js'
import { JSON_SCHEMA_ORDER } from './jsonSchema.cnst.js'
import type {
  JsonSchema,
  JsonSchemaAllOf,
  JsonSchemaAny,
  JsonSchemaArray,
  JsonSchemaBoolean,
  JsonSchemaConst,
  JsonSchemaEnum,
  JsonSchemaNull,
  JsonSchemaNumber,
  JsonSchemaObject,
  JsonSchemaOneOf,
  JsonSchemaRef,
  JsonSchemaString,
  JsonSchemaTuple,
} from './jsonSchema.model.js'
import { mergeJsonSchemaObjects } from './jsonSchema.util.js'

/* eslint-disable id-blacklist, @typescript-eslint/explicit-module-boundary-types */

export interface JsonSchemaBuilder<T = unknown> {
  build: () => JsonSchema<T>
}

/**
 * Fluent (chainable) API to manually create Json Schemas.
 * Inspired by Joi and Zod.
 */
export const j = {
  any<T = unknown>() {
    return new JsonSchemaAnyBuilder<T, JsonSchemaAny<T>>({})
  },
  const<T extends string | number | boolean | null>(value: T) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaConst<T>>({
      const: value,
    })
  },
  null() {
    return new JsonSchemaAnyBuilder<null, JsonSchemaNull>({
      type: 'null',
    })
  },
  ref<T = unknown>($ref: string) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaRef<T>>({
      $ref,
    })
  },

  enum<const T extends readonly (string | number | boolean | null)[] | StringEnum | NumberEnum>(
    values: T,
  ) {
    let enumValues: readonly (string | number | boolean | null)[] | undefined

    if (Array.isArray(values)) {
      enumValues = values
    } else if (typeof values === 'object') {
      if (isNumberEnum(values)) {
        enumValues = _numberEnumValues(values)
      } else if (isStringEnum(values)) {
        enumValues = _stringEnumValues(values)
      }
    }

    if (!enumValues) throw new TypeError('Unsupported enum input')

    return new JsonSchemaAnyBuilder<
      T extends readonly (infer U)[]
        ? U
        : T extends StringEnum
          ? T[keyof T]
          : T extends NumberEnum
            ? T[keyof T]
            : never,
      JsonSchemaEnum<any>
    >({
      enum: enumValues as any[],
    })
  },

  boolean() {
    return new JsonSchemaAnyBuilder<boolean, JsonSchemaBoolean>({
      type: 'boolean',
    })
  },
  buffer() {
    return new JsonSchemaAnyBuilder<Buffer, JsonSchemaAny<Buffer>>({
      instanceof: 'Buffer',
    })
  },

  // number types
  number<T extends number = number>() {
    return new JsonSchemaNumberBuilder<T>()
  },
  integer<T extends number = number>() {
    return new JsonSchemaNumberBuilder<T>().integer()
  },

  // string types
  string<T extends string = string>() {
    return new JsonSchemaStringBuilder<T>()
  },

  // complex types
  object,
  dbEntity<T extends AnyObject>(props: T) {
    return j
      .object<BaseDBEntity>({
        id: j.string(),
        created: j.integer().unixTimestamp2000(),
        updated: j.integer().unixTimestamp2000(),
      })
      .extend(j.object(props))
  },

  rootObject<T extends AnyObject>(props: {
    [K in keyof T]: JsonSchemaAnyBuilder<T[K]>
  }) {
    return new JsonSchemaObjectBuilder<T>().addProperties(props).$schemaDraft7()
  },
  array<T extends JsonSchemaAnyBuilder<any>>(itemSchema: T) {
    return new JsonSchemaArrayBuilder<T['infer']>(itemSchema)
  },
  tuple<T extends any[] = unknown[]>(items: JsonSchemaAnyBuilder[]) {
    return new JsonSchemaTupleBuilder<T>(items)
  },
  oneOf<Builders extends JsonSchemaAnyBuilder<any, any, any>[]>(
    items: [...Builders],
  ): JsonSchemaAnyBuilder<
    Builders[number] extends JsonSchemaAnyBuilder<infer U, any, any> ? U : never,
    JsonSchemaOneOf<Builders[number] extends JsonSchemaAnyBuilder<infer U, any, any> ? U : never>
  > {
    return new JsonSchemaAnyBuilder({
      oneOf: items.map(b => b.build()),
    }) as any
  },
  allOf<T = unknown>(items: JsonSchemaAnyBuilder[]) {
    return new JsonSchemaAnyBuilder<T, JsonSchemaAllOf<T>>({
      allOf: items.map(b => b.build()),
    })
  },
}

export class JsonSchemaAnyBuilder<
  T = unknown,
  SCHEMA_TYPE extends JsonSchema<T> = JsonSchema<T>,
  Opt extends boolean = false,
> implements JsonSchemaBuilder<T>
{
  constructor(protected schema: SCHEMA_TYPE) {}

  /**
   * Used in ObjectBuilder to access schema.optionalProperty
   */
  getSchema(): SCHEMA_TYPE {
    return this.schema
  }

  $schema($schema: string): this {
    Object.assign(this.schema, { $schema })
    return this
  }

  $schemaDraft7(): this {
    this.$schema('http://json-schema.org/draft-07/schema#')
    return this
  }

  $id($id: string): this {
    Object.assign(this.schema, { $id })
    return this
  }

  title(title: string): this {
    Object.assign(this.schema, { title })
    return this
  }

  description(description: string): this {
    Object.assign(this.schema, { description })
    return this
  }

  deprecated(deprecated = true): this {
    Object.assign(this.schema, { deprecated })
    return this
  }

  type(type: string): this {
    Object.assign(this.schema, { type })
    return this
  }

  default(v: any): this {
    Object.assign(this.schema, { default: v })
    return this
  }

  instanceof(of: string): this {
    this.schema.instanceof = of
    return this
  }

  optional(): JsonSchemaAnyBuilder<T | undefined, JsonSchema<T | undefined>, true>
  optional(optional: true): JsonSchemaAnyBuilder<T | undefined, JsonSchema<T | undefined>, true>
  optional(
    optional: false,
  ): JsonSchemaAnyBuilder<Exclude<T, undefined>, JsonSchema<Exclude<T, undefined>>, false>
  optional(optional = true): JsonSchemaAnyBuilder<any, JsonSchema<any>, false> {
    if (optional) {
      this.schema.optionalField = true
    } else {
      this.schema.optionalField = undefined
    }
    return this
  }

  nullable(): JsonSchemaAnyBuilder<T | null, JsonSchema<T | null>, Opt> {
    return new JsonSchemaAnyBuilder<T | null, JsonSchema<T | null>, Opt>({
      anyOf: [this.build(), { type: 'null' }],
    })
  }

  /**
   * Produces a "clean schema object" without methods.
   * Same as if it would be JSON.stringified.
   */
  build(): SCHEMA_TYPE {
    return _sortObject(JSON.parse(JSON.stringify(this.schema)), JSON_SCHEMA_ORDER)
  }

  clone(): JsonSchemaAnyBuilder<T, SCHEMA_TYPE, Opt> {
    return new JsonSchemaAnyBuilder<T, SCHEMA_TYPE, Opt>(_deepCopy(this.schema))
  }

  /**
   * @experimental
   */
  infer!: T
}

export class JsonSchemaNumberBuilder<
  T extends number = number,
  Opt extends boolean = false,
> extends JsonSchemaAnyBuilder<T, JsonSchemaNumber<T>, Opt> {
  constructor() {
    super({
      type: 'number',
    })
  }

  integer(): this {
    Object.assign(this.schema, { type: 'integer' })
    return this
  }

  multipleOf(multipleOf: number): this {
    Object.assign(this.schema, { multipleOf })
    return this
  }

  min(minimum: number): this {
    Object.assign(this.schema, { minimum })
    return this
  }

  exclusiveMin(exclusiveMinimum: number): this {
    Object.assign(this.schema, { exclusiveMinimum })
    return this
  }

  max(maximum: number): this {
    Object.assign(this.schema, { maximum })
    return this
  }

  exclusiveMax(exclusiveMaximum: number): this {
    Object.assign(this.schema, { exclusiveMaximum })
    return this
  }

  /**
   * Both ranges are inclusive.
   */
  range(minimum: number, maximum: number): this {
    Object.assign(this.schema, { minimum, maximum })
    return this
  }

  format(format: string): this {
    Object.assign(this.schema, { format })
    return this
  }

  int32 = (): this => this.format('int32')
  int64 = (): this => this.format('int64')
  float = (): this => this.format('float')
  double = (): this => this.format('double')

  unixTimestamp = (): JsonSchemaNumberBuilder<UnixTimestamp> =>
    this.integer().branded<UnixTimestamp>().format('unixTimestamp').description('UnixTimestamp')

  unixTimestamp2000 = (): JsonSchemaNumberBuilder<UnixTimestamp> =>
    this.integer()
      .branded<UnixTimestamp>()
      .format('unixTimestamp2000')
      .description('UnixTimestamp2000')

  unixTimestampMillis = (): JsonSchemaNumberBuilder<UnixTimestampMillis> =>
    this.integer()
      .branded<UnixTimestampMillis>()
      .format('unixTimestampMillis')
      .description('UnixTimestampMillis')

  unixTimestampMillis2000 = (): JsonSchemaNumberBuilder<UnixTimestampMillis> =>
    this.integer()
      .branded<UnixTimestampMillis>()
      .format('unixTimestampMillis2000')
      .description('UnixTimestampMillis2000')

  utcOffset = (): this => this.format('utcOffset')
  utcOffsetHours = (): this => this.format('utcOffsetHours')

  branded<B extends number>(): JsonSchemaNumberBuilder<B> {
    return this as unknown as JsonSchemaNumberBuilder<B>
  }
}

export class JsonSchemaStringBuilder<
  T extends string = string,
  Opt extends boolean = false,
> extends JsonSchemaAnyBuilder<T, JsonSchemaString<T>, Opt> {
  constructor() {
    super({
      type: 'string',
    })
  }

  regex(pattern: RegExp): this {
    return this.pattern(pattern.source)
  }

  pattern(pattern: string): this {
    Object.assign(this.schema, { pattern })
    return this
  }

  min(minLength: number): this {
    Object.assign(this.schema, { minLength })
    return this
  }

  max(maxLength: number): this {
    Object.assign(this.schema, { maxLength })
    return this
  }

  length(minLength: number, maxLength: number): this {
    Object.assign(this.schema, { minLength, maxLength })
    return this
  }

  format(format: string): this {
    Object.assign(this.schema, { format })
    return this
  }

  email = (): this => this.format('email')
  url = (): this => this.format('url')
  ipv4 = (): this => this.format('ipv4')
  ipv6 = (): this => this.format('ipv6')
  password = (): this => this.format('password')
  id = (): this => this.format('id')
  slug = (): this => this.format('slug')
  semVer = (): this => this.format('semVer')
  languageTag = (): this => this.format('languageTag')
  countryCode = (): this => this.format('countryCode')
  currency = (): this => this.format('currency')

  trim = (trim = true): this => this.transformModify('trim', trim)
  toLowerCase = (toLowerCase = true): this => this.transformModify('toLowerCase', toLowerCase)
  toUpperCase = (toUpperCase = true): this => this.transformModify('toUpperCase', toUpperCase)

  branded<B extends string>(): JsonSchemaStringBuilder<B> {
    return this as unknown as JsonSchemaStringBuilder<B>
  }

  /**
   * Accepts only the `YYYY-MM-DD` shape from all ISO 8601 variants.
   */
  isoDate(): JsonSchemaStringBuilder<IsoDate> {
    return this.format('IsoDate').branded<IsoDate>().description('IsoDate')
  }

  /**
   * Accepts strings that start with the `YYYY-MM-DDTHH:MM:SS` shape
   * and optionally end with either a `Z` or a `+/-hh:mm` timezone part.
   */
  isoDateTime(): JsonSchemaStringBuilder<IsoDateTime> {
    return this.format('IsoDateTime').branded<IsoDateTime>().description('IsoDateTime')
  }

  jwt(): this {
    return this.regex(JWT_REGEX)
  }

  private transformModify(t: 'trim' | 'toLowerCase' | 'toUpperCase', add: boolean): this {
    if (add) {
      this.schema.transform = _uniq([...(this.schema.transform || []), t])
    } else {
      this.schema.transform = this.schema.transform?.filter(s => s !== t)
    }
    return this
  }

  // contentMediaType?: string
  // contentEncoding?: string
}

export class JsonSchemaObjectBuilder<
  T extends AnyObject,
  Opt extends boolean = false,
> extends JsonSchemaAnyBuilder<T, JsonSchemaObject<T>, Opt> {
  constructor() {
    super({
      type: 'object',
      properties: {} as T,
      required: [],
      additionalProperties: false,
    })
  }

  addProperties(props: { [k in keyof T]: JsonSchemaBuilder<T[k]> }): this {
    Object.entries(props).forEach(([k, builder]: [keyof T, JsonSchemaBuilder]) => {
      const schema = builder.build()
      if (!schema.optionalField) {
        this.schema.required.push(k)
      } else {
        schema.optionalField = undefined
      }
      this.schema.properties[k] = schema
    })

    this.required(this.schema.required) // ensure it's sorted and _uniq

    return this
  }

  /**
   * Ensures `required` is always sorted and _uniq
   */
  required(required: (keyof T)[]): this {
    Object.assign(this.schema, { required })
    this.schema.required = _uniq(required).sort()
    return this
  }

  addRequired(required: (keyof T)[]): this {
    return this.required([...this.schema.required, ...required])
  }

  minProps(minProperties: number): this {
    Object.assign(this.schema, { minProperties })
    return this
  }

  maxProps(maxProperties: number): this {
    Object.assign(this.schema, { maxProperties })
    return this
  }

  additionalProps(additionalProperties: boolean): this {
    Object.assign(this.schema, { additionalProperties })
    return this
  }

  baseDBEntity(): JsonSchemaObjectBuilder<T & BaseDBEntity> {
    Object.assign(this.schema.properties, {
      id: { type: 'string' },
      created: { type: 'number', format: 'unixTimestamp2000' },
      updated: { type: 'number', format: 'unixTimestamp2000' },
    })

    return this.addRequired(['id', 'created', 'updated']) as any
  }

  extend<T2 extends AnyObject>(
    s2: JsonSchemaObjectBuilder<T2>,
  ): JsonSchemaObjectBuilder<T & T2 extends infer O ? { [K in keyof O]: O[K] } : never> {
    const builder = new JsonSchemaObjectBuilder<any>()
    Object.assign(builder.schema, _deepCopy(this.schema))
    mergeJsonSchemaObjects(builder.schema, s2.schema)
    return builder
  }
}

export class JsonSchemaArrayBuilder<ITEM, Opt extends boolean = false> extends JsonSchemaAnyBuilder<
  ITEM[],
  JsonSchemaArray<ITEM>,
  Opt
> {
  constructor(itemsSchema: JsonSchemaBuilder<ITEM>) {
    super({
      type: 'array',
      items: itemsSchema.build(),
    })
  }

  min(minItems: number): this {
    Object.assign(this.schema, { minItems })
    return this
  }

  max(maxItems: number): this {
    Object.assign(this.schema, { maxItems })
    return this
  }

  unique(uniqueItems: number): this {
    Object.assign(this.schema, { uniqueItems })
    return this
  }
}

export class JsonSchemaTupleBuilder<T extends any[]> extends JsonSchemaAnyBuilder<
  T,
  JsonSchemaTuple<T>
> {
  constructor(items: JsonSchemaBuilder[]) {
    super({
      type: 'array',
      items: items.map(b => b.build()),
      minItems: items.length,
      maxItems: items.length,
    })
  }
}

function object<P extends Record<string, JsonSchemaAnyBuilder<any, any, any>>>(
  props: P,
): JsonSchemaObjectBuilder<
  {
    [K in keyof P as P[K] extends JsonSchemaAnyBuilder<any, any, infer Opt>
      ? Opt extends true
        ? never
        : K
      : never]: P[K] extends JsonSchemaAnyBuilder<infer U, any, any> ? U : never
  } & {
    [K in keyof P as P[K] extends JsonSchemaAnyBuilder<any, any, infer Opt>
      ? Opt extends true
        ? K
        : never
      : never]?: P[K] extends JsonSchemaAnyBuilder<infer U, any, any> ? U : never
  } extends infer O
    ? { [K in keyof O]: O[K] }
    : never
>
function object<T extends AnyObject>(props: {
  [K in keyof T]: JsonSchemaAnyBuilder<T[K]>
}): JsonSchemaObjectBuilder<T>

function object(props: any): any {
  return new JsonSchemaObjectBuilder<any>().addProperties(props)
}
