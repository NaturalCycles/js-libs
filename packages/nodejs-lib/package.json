{
  "name": "@naturalcycles/nodejs-lib",
  "type": "module",
  "version": "14.3.0",
  "scripts": {
    "build": "dev-lib build",
    "test": "dev-lib test",
    "lint": "dev-lib lint",
    "bt": "dev-lib bt",
    "clean": "dev-lib clean",
    "check": "dev-lib check",
    "typecheck": "dev-lib typecheck",
    "slack-this-debug": "tsx ./src/bin/slack-this.ts --channel test --msg 'Hello slack!'",
    "secrets-gen-key-debug": "tsx ./src/bin/secrets-gen-key.ts",
    "secrets-encrypt-debug": "tsx ./src/bin/secrets-encrypt.ts",
    "secrets-decrypt-debug": "tsx ./src/bin/secrets-decrypt.ts",
    "kpy-debug": "tsx ./src/bin/kpy.ts --verbose scripts tmp/scripts",
    "kpy-debug2": "tsx ./src/bin/kpy.ts --verbose scripts bench non-ex non-ex/** colors* tmp/scripts",
    "kpy-debug3": "tsx ./src/bin/kpy.ts --verbose src colors csv stream non-ex non-ex/** tmp/src",
    "json2env-debug": "tsx ./src/bin/json2env.ts ./src/test/someFile.json"
  },
  "dependencies": {
    "@naturalcycles/js-lib": "workspace:^15",
    "@types/js-yaml": "^4",
    "@types/jsonwebtoken": "^9",
    "@types/yargs": "^16",
    "ajv": "^8",
    "ajv-formats": "^3",
    "chalk": "^5",
    "dotenv": "^17",
    "joi": "^18",
    "js-yaml": "^4",
    "jsonwebtoken": "^9",
    "lru-cache": "^11",
    "through2-concurrent": "^2",
    "tinyglobby": "^0.2",
    "tslib": "^2",
    "yargs": "^18"
  },
  "devDependencies": {
    "@naturalcycles/dev-lib": "workspace:*",
    "@types/through2-concurrent": "^2"
  },
  "exports": {
    ".": "./dist/index.js",
    "./lruMemoCache": "./dist/cache/lruMemoCache.js",
    "./colors": "./dist/colors/colors.js",
    "./csv": "./dist/csv/index.js",
    "./exec2": "./dist/exec2/exec2.js",
    "./fs2": "./dist/fs/fs2.js",
    "./env": "./dist/fs/json2env.js",
    "./kpy": "./dist/fs/kpy.js",
    "./yaml2": "./dist/fs/yaml2.js",
    "./glob": "./dist/glob/index.js",
    "./jwt": "./dist/jwt/jwt.service.js",
    "./runScript": "./dist/script/runScript.js",
    "./slack": "./dist/slack/index.js",
    "./stream": "./dist/stream/index.js",
    "./stream/*.js": "./dist/stream/*.js",
    "./yargs": "./dist/yargs/yargs.util.js",
    "./ajv": "./dist/validation/ajv/index.js",
    "./joi": "./dist/validation/joi/index.js",
    "./zip": "./dist/zip/zip.util.js"
  },
  "bin": {
    "kpy": "dist/bin/kpy.js",
    "json2env": "dist/bin/json2env.js",
    "generate-build-info": "dist/bin/generate-build-info.js",
    "slack-this": "dist/bin/slack-this.js",
    "secrets-gen-key": "dist/bin/secrets-gen-key.js",
    "secrets-encrypt": "dist/bin/secrets-encrypt.js",
    "secrets-decrypt": "dist/bin/secrets-decrypt.js"
  },
  "files": [
    "dist",
    "src",
    "!src/test",
    "!src/**/*.test.ts",
    "!src/**/__snapshots__",
    "!src/**/__exclude"
  ],
  "types": "dist/index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/NaturalCycles/js-libs.git",
    "directory": "packages/nodejs-lib"
  },
  "engines": {
    "node": ">=22.12.0"
  },
  "description": "Standard library for Node.js",
  "author": "Natural Cycles Team",
  "license": "MIT",
  "envByBranch": {
    "master": "master",
    "main": "master",
    "*": "branch"
  }
}
