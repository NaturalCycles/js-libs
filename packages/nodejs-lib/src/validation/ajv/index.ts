import Ajv from 'ajv'

export * from './ajvSchema.js'
export * from './ajvValidationError.js'
export * from './getAjv.js'
export * from './jsonSchemaBuilder.js'

export { Ajv }
