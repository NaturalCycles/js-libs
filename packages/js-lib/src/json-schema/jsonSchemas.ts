import type { BaseDBEntity } from '../types.js'
import { j } from './jsonSchemaBuilder.js'

export const baseDBEntityJsonSchema = j.object<BaseDBEntity>({
  id: j.string(),
  created: j.unixTimestamp2000(),
  updated: j.unixTimestamp2000(),
})
