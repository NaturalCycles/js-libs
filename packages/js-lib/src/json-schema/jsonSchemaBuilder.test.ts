import { AjvSchema } from '@naturalcycles/nodejs-lib/ajv'
import { describe, expect, expectTypeOf, test } from 'vitest'
import { _stringify } from '../string/stringify.js'
import type { BaseDBEntity, Branded, IsoDate, UnixTimestamp } from '../types.js'
import { z } from '../zod/index.js'
import { j } from './jsonSchemaBuilder.js'
import { baseDBEntityJsonSchema } from './jsonSchemas.js'

interface Address {
  createdAt: UnixTimestamp
  createdDate: IsoDate
  countryCode: string
  zip: string
  city: string
  address1: string
  address2?: string
  region?: string
  phone?: string
}

// interface AddressBM extends Address, BaseDBEntity {}

const addressJsonSchema = j.rootObject<Address>({
  createdAt: j.unixTimestamp2000(),
  createdDate: j.isoDate(),
  countryCode: j.string().countryCode(),
  zip: j.string().length(1, 40),
  city: j.string(),
  address1: j.string(),
  address2: j.string().optional(),
  region: j.string().optional(),
  phone: j.string().optional(),
})

const addressZodSchema = z.object({
  createdAt: z.unixTimestamp2000(),
  createdDate: z.isoDate(),
  countryCode: z.string(),
  zip: z.string().min(1).max(40),
  city: z.string(),
  address1: z.string(),
  address2: z.string().optional(),
  region: z.string().optional(),
  phone: z.string().optional(),
})

const addressBMJsonSchema = addressJsonSchema.extend(baseDBEntityJsonSchema)

// alternative

const addressBMJsonSchema2 = j
  .rootObject({})
  .extend(baseDBEntityJsonSchema.extend(addressJsonSchema))

// alternative 2
const addressBMJsonSchema3 = addressJsonSchema.extend(
  j.object<BaseDBEntity>({
    id: j.string(),
    created: j.unixTimestamp2000(),
    updated: j.unixTimestamp2000(),
  }),
)

test('simpleStringSchema', () => {
  const s = j.string().build()

  expect(s).toMatchInlineSnapshot(`
    {
      "type": "string",
    }
  `)
})

test('addressSchema', () => {
  expect(addressJsonSchema.build()).toMatchSnapshot()
})

test('addressBMJsonSchema', () => {
  expect(addressBMJsonSchema.build()).toMatchSnapshot()
  expect(addressBMJsonSchema2.build()).toEqual(addressBMJsonSchema.build())
  expect(addressBMJsonSchema3.build()).toEqual(addressBMJsonSchema.build())
})

test('oneOf', () => {
  const s = j.allOf([j.string(), j.string().countryCode()])
  expect(s.build()).toMatchInlineSnapshot(`
    {
      "allOf": [
        {
          "type": "string",
        },
        {
          "format": "countryCode",
          "type": "string",
        },
      ],
    }
  `)
})

test('order', () => {
  const s = addressBMJsonSchema.$schemaDraft7().$id('AddressBM').build()
  expect(Object.keys(s)).toMatchInlineSnapshot(`
    [
      "$schema",
      "$id",
      "type",
      "properties",
      "required",
      "additionalProperties",
    ]
  `)
})

test('buffer', () => {
  const s = j.buffer()
  expect(s.build()).toMatchInlineSnapshot(`
    {
      "instanceof": "Buffer",
    }
  `)

  // const schema = AjvSchema.create(s) // this fails strangely!
  const schema = AjvSchema.create(s.build())
  schema.validate(Buffer.from('abc'))

  expect(schema.isValid('a b c' as any)).toBe(false)
})

// todo: there are still differences, let's review them
test.todo('compare with zod json schema', () => {
  const zodJsonSchema = z.toJSONSchema(addressZodSchema, { target: 'draft-7' })
  const jsonSchema = addressJsonSchema.build()

  zodJsonSchema.required!.sort() // for snapshot stability

  expect(jsonSchema).toEqual(zodJsonSchema)
})

describe('integer', () => {
  describe('branded', () => {
    type BPM = Branded<number, 'BPM'>

    test('should assign the branded type', () => {
      const schema = j.object({
        foo: j.number().branded<BPM>(),
      })
      const [, result] = AjvSchema.create(schema.build()).getValidationResult({ foo: 1 as BPM })
      result satisfies { foo: BPM }
    })

    test('should be chainable with other commands', () => {
      const schema = j.object({
        foo: j.number().branded<BPM>().max(2),
      })

      const [err, result] = AjvSchema.create(schema.build()).getValidationResult({
        foo: 10 as BPM,
      })

      expect(err).toMatchInlineSnapshot(`
        [AjvValidationError: Object/foo must be <= 2
        Input: { foo: 10 }]
      `)
      expect(result).toEqual({ foo: 10 })
      expectTypeOf(result).toEqualTypeOf<{ foo: BPM }>()
    })
  })
})

describe('string', () => {
  describe('branded', () => {
    type AccountId = Branded<string, 'AccountId'>

    test('should assign the branded type', () => {
      const schema = j.object({
        foo: j.string().branded<AccountId>(),
      })
      const [, result] = AjvSchema.create(schema.build()).getValidationResult({
        foo: 'bingbong' as AccountId,
      })
      expectTypeOf(result).toEqualTypeOf<{ foo: AccountId }>()
    })

    test('should be chainable with other commands', () => {
      const schema = j.object({
        foo: j.string().branded<AccountId>().max(2),
      })

      const [err, result] = AjvSchema.create(schema.build()).getValidationResult({
        foo: 'bingbong' as AccountId,
      })

      expect(err).toMatchInlineSnapshot(`
        [AjvValidationError: Object/foo must NOT have more than 2 characters
        Input: { foo: 'bingbong' }]
      `)
      expect(result).toEqual({ foo: 'bingbong' })
      expectTypeOf(result).toEqualTypeOf<{ foo: AccountId }>()
    })
  })
})

describe('array', () => {
  test('should correctly infer the type from the type of the items', () => {
    const schema = j.object({ foo: j.array(j.number()) })
    const [, result] = AjvSchema.create(schema.build()).getValidationResult({ foo: [1, 2, 3] })
    result satisfies { foo: number[] }
    expectTypeOf(result).toEqualTypeOf<{ foo: number[] }>()
  })
})

describe('optional', () => {
  test('should correctly infer the type of optional fields', () => {
    interface Foo {
      reqNum: number
      optNum?: number
      reqStr: string
      optStr?: string
      reqArrOfReqNum: number[]
      optArrOfReqNum?: number[] | undefined
    }
    const schema = j.object({
      reqNum: j.number(),
      optNum: j.number().optional(),
      reqStr: j.string(),
      optStr: j.string().optional(),
      reqArrOfReqNum: j.array(j.number()),
      optArrOfReqNum: j.array(j.number()).optional(),
    })
    const badSchema = j.object({
      reqNum: j.number().optional(),
      optNum: j.number(),
      reqStr: j.string().optional(),
      optStr: j.string(),
    })

    const [, result] = AjvSchema.create(schema.build()).getValidationResult({} as any)

    result satisfies Foo

    const [, resultOfBadSchema] = AjvSchema.create(badSchema.build()).getValidationResult({} as any)

    // @ts-expect-error
    resultOfBadSchema satisfies Foo
  })

  test('should correctly require an optional property', () => {
    interface Foo {
      reqProp: string
    }
    const optionalString = j.string().optional()
    const schema = j.object({
      reqProp: optionalString.optional(false),
    })

    const [err, result] = AjvSchema.create(schema.build()).getValidationResult({} as any)
    result satisfies Foo
    expect(err).not.toBeNull()
  })

  test('should mark the property as optional', () => {
    const schema = j.object({
      foo: j.string().optional(),
      bar: j.boolean().optional(),
    })
    const jsonSchema = schema.build()
    const testCases = [{} as any, { foo: undefined }, { foo: 'bar' }, { foo: 'bar', bar: true }]

    testCases.forEach(test => {
      const [err] = AjvSchema.create(jsonSchema).getValidationResult(test)
      // eslint-disable-next-line vitest/valid-expect
      expect(err, _stringify(test)).toBeNull()
    })
  })
})
