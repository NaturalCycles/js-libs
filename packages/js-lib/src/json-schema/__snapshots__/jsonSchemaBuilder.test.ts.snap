// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addressBMJsonSchema 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "address1": {
      "type": "string",
    },
    "address2": {
      "type": "string",
    },
    "city": {
      "type": "string",
    },
    "countryCode": {
      "format": "countryCode",
      "type": "string",
    },
    "created": {
      "description": "UnixTimestamp2000",
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "createdAt": {
      "description": "UnixTimestamp2000",
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "createdDate": {
      "description": "IsoDate",
      "format": "isoDate",
      "type": "string",
    },
    "id": {
      "type": "string",
    },
    "phone": {
      "type": "string",
    },
    "region": {
      "type": "string",
    },
    "updated": {
      "description": "UnixTimestamp2000",
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "zip": {
      "maxLength": 40,
      "minLength": 1,
      "type": "string",
    },
  },
  "required": [
    "address1",
    "city",
    "countryCode",
    "created",
    "createdAt",
    "createdDate",
    "id",
    "updated",
    "zip",
  ],
  "type": "object",
}
`;

exports[`addressSchema 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "properties": {
    "address1": {
      "type": "string",
    },
    "address2": {
      "type": "string",
    },
    "city": {
      "type": "string",
    },
    "countryCode": {
      "format": "countryCode",
      "type": "string",
    },
    "createdAt": {
      "description": "UnixTimestamp2000",
      "format": "unixTimestamp2000",
      "type": "number",
    },
    "createdDate": {
      "description": "IsoDate",
      "format": "isoDate",
      "type": "string",
    },
    "phone": {
      "type": "string",
    },
    "region": {
      "type": "string",
    },
    "zip": {
      "maxLength": 40,
      "minLength": 1,
      "type": "string",
    },
  },
  "required": [
    "address1",
    "city",
    "countryCode",
    "createdAt",
    "createdDate",
    "zip",
  ],
  "type": "object",
}
`;
